370-auth-ui/sdk.md

```diff
@@ -0,0 +1,306 @@
+    </View>
+  );
+}
+```
+
+### Browser-based App SDK
+
+#### API
+
+```typescript
+interface AuthorizeOptions {
+  redirectURI: string;
+  state?: string;
+  onUserDuplicate?: "abort" | "merge" | "create";
+}
+
+// Modeled after https://tools.ietf.org/html/rfc6749#section-5.2
+interface OAuthError {
+  error: string;
+  error_description?: string;
+  error_uri?: string;
+}
+
+// The promise resolves to nothing. It is because after calling this function,
+// the URL is changed. The app is destoryed so the return value is not usable at all.
+function authorize(options: AuthorizeOptions): Promise<void>;
+
+// exchangeToken looks at window.location
+// It checks if error is present and rejects with OAuthError.
+// Otherwise if code is present, make a token request.
+function exchangeToken(): Promise<User>;
+```
+
+#### Example
+
+##### Authenticate in a Browser-based application
+
+```typescript
+import React, { useCallback, useEffect } from "react";
+import skygear from "@skygear/web";
+
+function AuthenticationScreen() {
+  const onClick = useCallback(() => {
+    await skygear.auth.authorizeWithWeb({
+      redirectURI: "https://www.myapp.com/continue",
+      state: "mystate",
+    });
+  }, []);
+  return (
+    <div>
+      <button onClick={onClick}>Sign in</button>
+    </div>
+  );
+}
+
+function ContinueScreen() {
+  useEffect(() => {
+    try {
+      const u = new URL(window.location.href);
+      const state = u.searchParams.get("state");
+
+      const user = await skygear.auth.exchangeToken();
+      if (!user.isVerified) {
+        window.location = "https://myapp.skygearapps.com/verify";
+      } else {
+        window.location = "https://myapp.skygearapps.com/home";
+      }
+    } catch (e) {
+      // Handle the error properly here.
+      // For example, duplicated user.
+      // Or
+      // This route is likely to be accessed manually.
+      // Redirect to home
+      window.location = "https://myapp.skygearapps.com/home";
+    }
+  }, []);
+  return (
+    <div>Loading</div>
+  );
+}
+```
+
+##### Open settings in a Browser-based application
+
+```typescript
+import React, { useCallback } from "react";
+import skygear from "@skygear/web";
+
+function SettingsScreen() {
+  const url = "https://accounts.myapp.com";
+  const onClickOpenSettingsAsPopup = useCallback(() => {
+    window.open(url);
+  }, []);
+  const onClickOpenSettings = useCallback(() => {
+    window.location = url;
+  }, []);
+  return (
+    <div>
+      <button onClick={onClickOpenSettingsAsPopup}>Open Settings as popup</button>
+      <button onClick={onClickOpenSettings}>Open Settings</button>
+      <iframe src={url} />
+    </div>
+  );
+}
+```
+
+## The open source AppAuth library
+
+The open source AppAuth library has three variants, namely Android, iOS and JS. The JS one is for web. However, our JS SDK must support three common JS environments browser, Node and React Native. Therefore AppAuth is not being used.
```

Comment:
I think AuthAuth for JS can be used in Node too? Quoting from their README:
> The library is designed for use in `Web Apps`, `Node.js` CLI applications,
`Chrome Apps` and applications that use `Electron` or similar frameworks.

Reply:
But not React Native :(

Reply:
I found this: https://github.com/FormidableLabs/react-native-app-auth

Is it suitable?

Reply:
The AppAuth library is not particularly useful in our use case. The most challenge part of the SDK IMO is the refresh flows, which are not covered by AppAuth. The PKCE utilities are already implemented in our SDK. POSTing form request can be done with `fetch` easily.

Reply:
Okay I understand. Then, we can just note down there exists this popular official library, and we may want to test our server-side implementation against it.